lib/go-parser/version-parser.ts

```diff
@@ -0,0 +1,196 @@
+import { Elf, ElfProgram, GoModulesResult, ReadPtrFunc } from "./types";
+
+/**
+ * Create same output as `go version -m binary-file` does
+ * @param binary
+ */
+export function extractModulesFromBinary(binary: Elf): GoModulesResult {
+  const { version: goVersion, mod } = findVers(binary);
+  const { name, modules } = prepareGoDependencies(mod);
+
+  return { goVersion, name, modules };
+}
+
+/**
+ * Normalize versions to align with `snyk-go-parser`
+ * @param mod
+ */
+function prepareGoDependencies(mod: string) {
+  if (!mod) {
+    return { name: "", modules: {} };
+  }
+
+  const [, mainModuleLine, ...versionsLines] = mod.split("\n");
+  const [, name] = mainModuleLine.split("\t");
+  const modules = versionsLines.reduce((accum, curr) => {
+    // Skip If for some reason after splitting there is a empty line
+    if (!curr) {
+      return accum;
+    }
+
+    const [, name, ver] = curr.split("\t");
+    // Versions in Go have trailing 'v'
+    let version = ver.substring(1);
+    // In versions with hash, we only care about hash
+    // v0.0.0-20200905004654-be1d3432aa8f => #be1d3432aa8f
+    version = version.includes("-")
+      ? `#${version.substring(version.lastIndexOf("-") + 1)}`
+      : version;
+
+    accum[name] = version;
+    return accum;
+  }, {});
+
+  return { name, modules };
+}
+
+/**
+ * Function finds and returns the Go version and
+ * module version information in the executable binary
+ * @param binary
+ */
+function findVers(binary: Elf) {
+  const result = {
+    version: "",
+    mod: "",
+  };
+
+  const text = dataStart(binary);
+  let data = readData(binary.body.programs, text, 64 * 1024) || Buffer.from([]);
+
+  for (
+    ;
+    !data.toString("binary").startsWith("\xff Go buildinf:");
+    data = data.slice(32)
+  ) {
+    if (data.length < 32) {
+      return result;
+    }
+  }
+
+  const ptrSize = data[14];
+  const bigEndian = data[15] !== 0;
+
+  let readPtr: ReadPtrFunc;
+  if (ptrSize === 4) {
+    if (bigEndian) {
+      readPtr = (buffer) => buffer.readInt32BE(0);
+    } else {
+      readPtr = (buffer) => buffer.readInt32LE(0);
+    }
+  } else {
+    if (bigEndian) {
+      readPtr = (buffer) => Number(buffer.readBigInt64BE());
+    } else {
+      readPtr = (buffer) => Number(buffer.readBigInt64LE());
+    }
+  }
+
+  const version: string = readString(
+    binary,
+    ptrSize,
+    readPtr,
+    readPtr(data.slice(16, 16 + ptrSize)),
+  );
+
+  if (version === "") {
+    return result;
+  }
+
+  result.version = version;
+
+  const mod: string = readString(
+    binary,
+    ptrSize,
+    readPtr,
+    readPtr(data.slice(16 + ptrSize, 16 + 2 * ptrSize)),
+  );
+
+  if (mod.length >= 33 && mod[mod.length - 17] === "\n") {
+    result.mod = mod.slice(16, mod.length - 16);
+  } else {
```

Comment:
Should we describe what this is trying to do?

Reply:
Done :)